chain/chain/src/state_transition_data.rs

```diff
@@ -0,0 +1,265 @@
+use near_store::{DBCol, StorageError};
+use once_cell::unsync::OnceCell;
+
+use crate::{Chain, ChainStore, ChainStoreAccess};
+
+#[derive(BorshSerialize, BorshDeserialize, PartialEq, Eq, Default, Debug)]
+struct StateTransitionsMetadata {
+    /// Represents max not-yet-GCed height for each shard
+    start_heights: HashMap<ShardId, BlockHeight>,
+}
+
+impl Chain {
+    pub(crate) fn garbage_collect_state_transition_data(&self, block: &Block) -> Result<(), Error> {
+        let protocol_version =
+            self.epoch_manager.get_epoch_protocol_version(block.header().epoch_id())?;
+        if cfg!(feature = "shadow_chunk_validation")
+            || checked_feature!("stable", StatelessValidationV0, protocol_version)
+        {
+            let chain_store = self.chain_store();
+            let final_block_hash = *block.header().last_final_block();
+            if final_block_hash == CryptoHash::default() {
+                return Ok(());
+            }
+            let final_block = chain_store.get_block(&final_block_hash)?;
+            let final_block_height = final_block.header().height();
+            let final_block_present_chunks = final_block
+                .chunks()
+                .iter()
+                .filter(|chunk| chunk.is_new_chunk(final_block_height))
+                .map(|chunk| chunk.shard_id())
+                .collect::<Vec<_>>();
+            clear_before_last_final_block(
+                chain_store,
+                final_block_height,
+                &final_block_present_chunks,
+            )?;
+        }
+        Ok(())
+    }
+}
+
+/// Removes StateTransitionData entries from the db based on last final block state.
+/// It is safe to delete all StateTransitionData before the last final block
+/// for the shards with present chunks in that block.
+/// TODO(resharding): this doesn't work after shard layout change
+fn clear_before_last_final_block(
+    chain_store: &ChainStore,
+    last_final_block_height: BlockHeight,
+    last_final_block_present_chunks: &[ShardId],
+) -> Result<(), Error> {
+    let mut metadata = chain_store
+        .store()
+        .get_ser::<StateTransitionsMetadata>(DBCol::Misc, STATE_TRANSITIONS_METADATA_KEY)?
+        .unwrap_or_default();
+    tracing::debug!(
+        target: "state_transition_data",
+        last_final_block_height,
+        ?metadata,
+        "garbage collecting state transition data"
+    );
+    // Only calculated if metadata doesn't exist or if it doesn't contain start_heights entry for the shard
+    let computed_start_heights = OnceCell::<HashMap<ShardId, BlockHeight>>::new();
+    let mut store_update = chain_store.store().store_update();
+    for &shard_id in last_final_block_present_chunks {
+        let start_height = if let Some(&start_height) = metadata.start_heights.get(&shard_id) {
+            start_height
+        } else {
+            *computed_start_heights
+                .get_or_try_init(|| compute_start_heights(chain_store))?
```

Comment:
Seems like the purpose of this is basically for the first time initialization? i.e. when we don't have the start_heights stored in metadata? Could you confirm ideally this should be called only once to initialize `STATE_TRANSITIONS_METADATA_KEY`?

Instead of calling `compute_start_heights` here, how about we just take the tail height from chain here? Basically the height at which we are currently doing GC?

Reply:
Yes, we only need it for the first initialisation when `start_heights` is not yet in the db. Now it is easier to see that with d0b14fc1d310ac02343bdc54fa07e4cf97e3a26e.
Using tail height from chain is not really feasible since it will result in 5 epoch-length worth of deletes which could slow down block processing, so I'd rather not risk it.

Reply:
Just for my understanding, isn't that what we are currently doing with the first time initialization as well? `compute_start_heights` iterates through the whole `DBCol::StateTransitionData` and gets the min start_heights for each shard right? And we can potentially have entries from upto 5 epochs (or GC period)?

Reply:
Sounds like in either case we are taking this one time hit of iterating through a lot of entries in `DBCol::StateTransitionData` for first time initialization of `StateTransitionStartHeights`. Assuming this is to solve the issue we are seeing with shadow validation, is it fine to take this one time slowdown hit?

Reply:
Oh and now I'm a bit concerned about archival nodes that maybe don't clear the `DBCol::StateTransitionData` column? I'm guessing there are no such nodes yet, but having this PR in place before DBCol::StateTransitionData grows for archival nodes should do the trick?

Reply:
> iterates through the whole DBCol::StateTransitionData and gets the min start_heights for each shard right? And we can potentially have entries from upto 5 epochs (or GC period)?

This can only be the case for stateless net, which shouldn't be an issue. Also `StateTransitionData` instances are fairly small there. 
On mainnet it will be enabled for the first time along with stateless validation protocol upgrade, first time cleanup will only have to read a few entries. I even considered not storing `start_heights` at all and always deriving that from the db, but the issues that even a few `StateTransitionData` result in multiple megabytes of data read from the disk, so I'd rather not do that.
The same applies for the archival nodes.